Draft: Configuration and hot-reloading. (#10)
openmoq/moqx: design/hot-reloading.md

## Hot-Reloading Design

### Terminology

> - **Config reload**: Re-reading and applying configuration without stopping the process. May affect only new or all connections depending on parameter lifecycle.

Reviewer: Config syntax should be commented and clear on which are run-time reloadable vs restart required.  For example, see postgres config.

Reviewer 2: I think there's a tradeoff between structuring the config hierarchically, which maps to the mental model, vs structuring it by reloadability and prefer the former.

- **Graceful restart**: Replacing the process binary. Old process drains existing connections; new process accepts new ones.

### 1. File-Based Config Reload (SIGHUP)

> Sending `SIGHUP` to the relay process triggers config reload, it does never trigger graceful restart.

Reviewer: ```suggestion
Sending `SIGHUP` to the relay process triggers config reload only.
```

Often for drain and graceful restart, SIGUSR1 or similar would be used.

Author: The draining is, IMHO, more complicated. I spun that off to a separate issue: https://github.com/openmoq/o-rly/issues/15
<!-- Reviewable comment -OmEbwp54noW0rwNjLzG:bkju2if -->
<!-- Sent from Reviewable.io -->

Attempting to change static config triggers an error. 

### 2. Runtime API

The relay exposes an HTTP API (per [Relay Requirements 3.3.2](https://github.com/OpenMOQ/relay-requirements/blob/main/relay-requirements.md)) for programmatic configuration. This API is served on a separate listener (the management/control plane listener) and is authenticated via mTLS or API key.

```

> GET    /v1/config                    # Return current running config (JSON)

Reviewer 2: DO these return yaml or json or .. ?

Author: JSON. 

YAML is designed as JSON superset. 
So returning JSON for better machine support but accepting YAML for better readability (and comments!) makes sense in all places.

Reviewer: Conversion from YAML is 1:1 to JSON.   It's not the other way around, such as YAML anchors/references, comments, ... IMHO, from an API perspective, JSON is more common and preferred. Config syntax documentation, file configs, etc. are often YAML formatted.

Author: @TimEvens So do you want a wording change here, or are you okay with the current state? 
<!-- Reviewable comment -OmERjZQEmQybXdXVY9-:btrvgd3 -->
<!-- Sent from Reviewable.io -->

This is important because config drift is a common operational problem. Envoy solves this by treating xDS as the source of truth (no file). HAProxy's runtime API explicitly does not persist changes to disk, which operators frequently complain about. We should default to persisting.

### 3. Graceful Binary Restart

For deploying new binaries or making changes to static parameters, the relay supports graceful restart where the old process drains while a new process takes over.

**eBPF-based steering**: An eBPF program attached to the UDP socket steers packets for existing connection IDs to the old process and new connections to the new process.

> **Note**: We will likely need eBPF-based steering between I/O threads anyway for QUIC connection affinity (ensuring all packets for a given connection are processed by the same thread). Extending this to process-level steering during restarts is incremental work on top of that foundation. See [QUIC-LB](https://datatracker.ietf.org/doc/draft-ietf-quic-load-balancers/) and [Relay Requirements §3.7](https://github.com/OpenMOQ/relay-requirements/blob/main/relay-requirements.md) for the connection ID generation API requirement.

> **State handoff**: We explicitly do **not** attempt to transfer in-memory state (cache, subscription state) between old and new processes. This is complex, error-prone, and the correct approach for stateless horizontal scaling is to rely on upstream re-subscription and cache warming. GOAWAY gives clients the signal to reconnect to the new process.

Reviewer 2: The only one we may need to consider is in-memory cache, but perhaps that can be delegated to a shared-memory caching provider since it's inherently more complex.

Author: +1 on delegating to a shared-memory caching provider. 

Any handoff or sharing between binaries during restart/upgrade requires API/ABI compatibility, limiting the usefulness and coming with performance tradeoffs.

Reviewer 2: LGTM
<!-- Reviewable comment -OmtZOQv5aFSGxeYiWq7:bb6ywmx -->
<!-- Sent from Reviewable.io -->
